File: ts/src/day10.ts
import { DayBase } from './day-runner.js';

// https://adventofcode.com/2021/day/10

const cscore = [   3,  57, 1197, 25137 ]; // corrupt scores
const ascore = [   1,   2,    3,     4 ]; // autocomplete scores
const opens  = [ '(', '[',  '{',   '<' ];
const closes = [ ')', ']',  '}',   '>' ];

export class Day10 extends DayBase {
    constructor() {
        super(10);
    }

    public async run(): Promise<void> {
        const lines = await this.getInputLines();
        // const lines = [                     // expect P1: 26397; P2: 288957
        //     '[({(<(())[]>[[{[]{<()<>>',
        //     '[(()[<>])]({[<{<<[]>>(',
        //     '{([(<{}[<>[]}>{[]{[(<()>',
        //     '(((({<>}<{<{<>}{[]{[]{}',
        //     '[[<[([]))<([[{}[[()]]]',
        //     '[{[{({}]{}}([{[{{{}}([]',
        //     '{<[[]]>}<{[{[{[]{()[[[]',
        //     '[<(<(<(<{}))><([]([]()',
        //     '<{([([[(<>()){}]>(<<{{',
        //     '<{([{{}}[<[[[<>{}]]]>[]]',
        // ];

        let part1 = 0;
        let autocompletescores: number[] = []
        lines.forEach((line, index) => {
            let corrupt = false;
            let stack: string[] = [];
            for (let i = 0; i < line.length; i++) {
                let ch = line[i];
                if (opens.includes(ch)) {
                    stack.push(ch);
                } else {
                    let opener = stack.pop() || "";
                    let oidx = opens.indexOf(opener);
                    if (ch !== closes[oidx]) {
                        //console.log(`line ${index} expected ${closes[oidx]} but found ${ch}`);
                        corrupt = true;
                        part1 += cscore[closes.indexOf(ch)];
                        break;
                    }
                }
            }
            if (!corrupt) {
                //let completion = '';
                let score = 0;
                while (stack.length > 0) {
                    let opener = stack.pop() || "";
                    let oidx = opens.indexOf(opener)
                    //completion += closes[oidx];
                    score = (score * 5) + ascore[oidx];
                }
                //console.log('Completion', completion);
                autocompletescores.push(score);
            }
        });

        autocompletescores.sort((a,b) => +a - +b); // Otherwise JavaScript will sort as if it were strings. Marvellous.
        const mid = ((autocompletescores.length - 1) / 2);
        //console.log('Autocomplete scores', autocompletescores, mid);
        console.info(`Day ${this.day}`);
        console.info(`  Part 1:`, part1);
        console.info(`  Part 2:`, autocompletescores[mid]);
        console.info('');
    }
}

File: py/day10.py
import os

# https://adventofcode.com/2021/day/10

with open(os.path.dirname(os.path.realpath(__file__)) + "/../input/day10-input", "r") as f:
    input = [line.strip() for line in f.readlines()]

corruptscores = [3, 57, 1197, 25137]
opench = [ '(', '[',  '{',   '<' ]
closech = [ ')', ']',  '}',   '>' ]

part1 = 0
autocomp = []
for line in input:
    stack = []
    corrupt = False
    for ch in line:
        if ch in opench:
            stack.append(ch)
        else:
            lastopen = stack.pop()
            if closech[opench.index(lastopen)] != ch:
                corrupt = True
                part1 += corruptscores[closech.index(ch)]
                break

    if not corrupt:
        score = 0
        while stack:
            opener = stack.pop()
            score = (score * 5) + opench.index(opener) + 1 # autocomplete scores are [1,2,3,4]
        autocomp.append(score)

autocomp.sort()
mid = (len(autocomp) - 1) // 2 # according to the puzzle text len() is always going to be odd for the autocomplete items
print("Day 10")
print("  Part 1:", part1)
print("  Part 2:", autocomp[mid])